of_design_doc.md

```diff
@@ -64,35 +58,29 @@ T = @of(
 For cases where dimensions need to be specified at runtime:
 
 ```julia
-# Define type with symbolic dimensions
-MatrixType = of((
-    rows=of(Int; constant=true),    # Integer type preserved
-    cols=of(Int; constant=true),    # Integer type preserved
-    data=of(Array, :rows, :cols),   # Note: expressions like :(rows + 1) not yet supported
-))
-
-# Simpler with @of macro
+# Define type with symbolic dimensions using @of macro
 MatrixType = @of(
     rows=of(Int; constant=true),
     cols=of(Int; constant=true),
     data=of(Array, rows, cols),  # Direct references converted to symbols
 )
 # Note: Expressions like 'rows + 1' are not yet supported
 
-# Create concrete type for use with flatten/unflatten
-ConcreteType = of(MatrixType; rows=3, cols=4)
-data = (data=rand(3, 4),)
-ConcreteType(data)
-flat = flatten(ConcreteType, data)
+# Create concrete type by specifying constants
+ConcreteType = MatrixType(;rows=3, cols=4)
+# Can also validate data while concretizing
+ConcreteType = MatrixType(;rows=3, cols=4, data=rand(3, 4))  # Validates that data matches 3x4
+data_nt = (data=rand(3, 4),)
+flat = flatten(ConcreteType, data_nt)
```

Comment:
Note `ConcreteType` already has randomly initialised `data`, so `data_nt` is unnecessary. 
```suggestion
flat = flatten(ConcreteType)
```

Reply:
this doesn't feel quite right -- `ConcreteType` in your example should be a type, i.e. 
```julia
ConcreteType == @of(
data = of(Array, 3, 4)
)
```
`flatten` should take a source of values that conforms to the `of` type and returns a vector.

Reply:
at the moment, as of the last commit

```julia
ConcreteType = MatrixType(;rows=3, cols=4, data=rand(3, 4))  # Validates that data matches 3x4
```
will only check that the size of `data` is consistent with
```julia
MatrixType = @of(
    rows=of(Int; constant=true),
    cols=of(Int; constant=true),
    data=of(Array, rows, cols),  # Direct references converted to symbols
)
```
with
 `rows=3, cols=4`

Reply:
> will only check that the size of data is consistent 

That would be confusing. I suggest

```julia
x = MatrixType(;rows=3)  # this should error with a message constant `cols` is required but not provided. 
x = MatrixType(;rows=3, cols=4)  # default to MatrixType(;rows=3, cols=4, data = zero(3, 4))
x = MatrixType(;rows=3, cols=4, data = data) # 
```

This would always return an instance when calling `T(;kwargs...)` instead of another type. 
In addition, it requires the user to pass all constants when calling `T(;kwargs...)`, and default to zero if non-constant variables aren't provided explicitly.

Reply:
this is updated now